CMakeLists.txt

```diff
@@ -0,0 +1,448 @@
+# MPLAB_PIC32MEC14XX               - Compiller: MPLAB         Target: PIC32MEC14xx
+# MPLAB_PIC32MX                    - Compiller: MPLAB         Target: PIC32MX
+# MPLAB_PIC32MZ                    - Compiller: MPLAB         Target: PIC32MZ
+# MSVC_MINGW                       - Compiller: MSVC or MinGW Target: x86
+# OWATCOM_16BIT_DOS_FLSH186        - Compiller: Open Watcom   Target: 16 bit DOS Flsh186
+# OWATCOM_16BIT_DOS_PC             - Compiller: Open Watcom   Target: 16 bit DOS PC
+# PARADIGM_TERN_EE_LARGE           - Compiller: Paradigm      Target: Tern EE large
+# PARADIGM_TERN_EE_SMALL           - Compiller: Paradigm      Target: Tern EE small
+# RENESAS_RX100                    - Compiller: Renesas       Target: RX100
+# RENESAS_RX200                    - Compiller: Renesas       Target: RX200
+# RENESAS_RX600                    - Compiller: Renesas       Target: RX600
+# RENESAS_RX600_V2                 - Compiller: Renesas       Target: RX600 v2
+# RENESAS_RX700_V3_DPFPU           - Compiller: Renesas       Target: RX700 v3 with DPFPU
+# RENESAS_SH2A_FPU                 - Compiller: Renesas       Target: SH2A with FPU
+# ROWLEY_MSP430F449                - Compiller: Rowley        Target: MSP430F449
+# RVDS_ARM_CA9                     - Compiller: RVDS          Target: ARM Cortex-A9
+# RVDS_ARM_CM0                     - Compiller: RVDS          Target: ARM Cortex-M0
+# RVDS_ARM_CM3                     - Compiller: RVDS          Target: ARM Cortex-M3
+# RVDS_ARM_CM4_MPU                 - Compiller: RVDS          Target: ARM Cortex-M4 with MPU
+# RVDS_ARM_CM4F                    - Compiller: RVDS          Target: ARM Cortex-M4 with FPU
+# RVDS_ARM_CM7                     - Compiller: RVDS          Target: ARM Cortex-M7
+# RVDS_ARM7_LPC21XX                - Compiller: RVDS          Target: ARM7 LPC21xx
+# SDCC_CYGNAL                      - Compiller: SDCC          Target: Cygnal
+# SOFTUNE_MB91460                  - Compiller: Softune       Target: MB91460
+# SOFTUNE_MB96340                  - Compiller: Softune       Target: MB96340
+# TASKING_ARM_CM4F                 - Compiller: Tasking       Target: ARM Cortex-M4 with FPU
+# CDK_THEAD_CK802                  - Compiller: CDK           Target: T-head CK802
+# XCC_XTENSA                       - Compiller: XCC           Target: Xtensa
+# WIZC_PIC18                       - Compiller: WizC          Target: PIC18
+
+add_library(freertos_kernel STATIC
+    croutine.c
+    event_groups.c
+    list.c
+    queue.c
+    stream_buffer.c
+    tasks.c
+    timers.c
+    portable/Common/mpu_wrappers.c
+
+    # If FREERTOS_HEAP is digit between 1 .. 5 - it is heap number, otherwise - it is path to custom heap source file
+    $<IF:$<BOOL:$<FILTER:${FREERTOS_HEAP},EXCLUDE,^[1-5]$>>,${FREERTOS_HEAP},portable/MemMang/heap_${FREERTOS_HEAP}.c>
+
+    $<$<STREQUAL:${FREERTOS_PORT},BCC_16BIT_DOS_FLSH186>:portable/BCC/16BitDOS/common/portcomn.c;portable/BCC/16BitDOS/Flsh186/port.c>
```

Comment:
Hello @yhsb2k, thank you for addressing my earlier comments. What if we move all this into functions? Would it be 'cleaner'?

Reply:
Something like this:
[CMakeLists.txt](https://github.com/FreeRTOS/FreeRTOS-Kernel/files/7850400/CMakeLists.txt)
[portable.cmake.txt](https://github.com/FreeRTOS/FreeRTOS-Kernel/files/7850404/portable.cmake.txt)

Change portable.cmake.txt to portable.cmake since github won't allow me to upload a file with .cmake extension

Reply:
@kstribrnAmzn what do you think?

Reply:
Overall I like it. I have been thinking of how we could combine the brevity of [PR 438](https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/438) and the completeness of this PR.

Point for conversation: Maybe rather than returning the exact set of files, the `get_portable_source` function could return a path which is used with file globbing. I agree with @jonenz comment that the portable files should be stable. And we can put a disclaimer in the readme for folks who do add new port files.

Thoughts @AniruddhaKanhere and @yhsb2k

Reply:
I am happy with that too. However, since out ports are stable, I would argue that why not return all the files required instead of globbing them?
@yhsb2k can you try moving some parts into functions as I described above? Or would you like me to push those changes?

Reply:
I have moved port handling into separate cmake file like @AniruddhaKanhere suggested.
But I did it using freertos_kernel_port target to comply with modern CMake approach.